adr/0009-extension-structure.adoc

```diff
@@ -0,0 +1,265 @@
+
+Most extensions follow the basic runtime/deployment split, but the boundaries between API and internal logic have often blurred.
+Additionally, without clear package structure rules, many split packages have emerged.
+This situation introduces several issues:
+
+* It is unclear which classes are intended as public APIs versus internal implementations.
+* There is no clear boundary between what an extension consumer can depend on and what they should not use.
+* Some extensions mix SPI and implementation code, reducing modularity and increasing the chance of unwanted coupling.
+* The need for build item re-use or coordination between extensions leads to ad-hoc patterns.
+
+More recently, we introduced dev-only modules, further increasing complexity.
+
+To address these issues and prepare for long-term maintainability (and possibly Java Platform Module System (JPMS) adoption), we propose a more structured layout for extension modules.
+
+== Decision
+
+Extensions may adopt a standardized structure, composed of one mandatory and several optional modules, depending on their complexity and integration needs:
+
+* `runtime`: Contains the runtime implementation.
+It may expose public APIs if needed but should prefer delegation to the `runtime-api` module (See <<the-place-of-the-public-api>>).
+This module also contains the `recorders` and other runtime logic.
+* `runtime-api`: Contains public APIs intended to be consumed by other extensions or application code.
+This module should have minimal dependencies.
+Depending on an `runtime-api` module does not transitively include the full extension (i.e., it avoids pulling in runtime or deployment logic). See <<coupling-api-and-spi>> for more details.
+* `deployment`: Contains build-time logic, including `BuildSteps`.
+This module may define internal build items.
+* `deployment-spi`: Contains build-time APIs intended to be reused across multiple extensions for integration purposes.
+Extensions contributing to or using build items from other extensions should depend on `deployment-spi` modules.
+* `dev`: Contains runtime classes used exclusively in dev mode (e.g., for the Dev UI). This avoids shipping dev-only classes into production artifacts.
+* `codestart`: Contains the _Codestart_ templates for the extension. See https://quarkus.io/guides/extension-codestart[Codestart Guide] for more details.
+* `codegen`: Contains code generation utilities such as code pre-processors invoked at build time.
+
+Only the `runtime` module is mandatory; the others are optional based on the extension's needs.
+
+
+.New structure for Quarkus extensions
+[source,tree]
+----
+my-extension/
+├── runtime/
+├── runtime-api/
+├── deployment/
+├── deployment-spi/
+├── dev/
+├── codegen/
+└── codestart/
+----
+
+=== Additional Rules
+
+Module dependencies must follow strict rules:
+* `deployment` depends on `runtime`, and `deployment-spi` (if defined).
+* `runtime` depends on `runtime-api` (if defined).
+* `dev`, if defined.
```

Comment:
This one isn't clear as a rule. Typically, `runtime` would declare a conditional dependency on `dev` based on lunch mode `dev`. Whether `deployment` depends on `dev`, would be a choice, not a strict rule.